gpjax/sparse_gps.py

```diff
@@ -0,0 +1,213 @@
+        hyperparams = concat_dictionaries(self.prior.params,
+                                          {"likelihood": self.likelihood.params,
+                                          "q": self.q.params})
+        return hyperparams
+
+    def elbo(self, train_data: Dataset, transformations: Dict) -> Callable[[Array], Array]:
+
+        def elbo_fn(params: Dict, batch: Dataset) -> Array:
+            params = transform(params, transformations)
+            kl = self.prior_kl(params)
+            var_exp = self.variational_expectation(params, batch)
+
+            return jnp.sum(var_exp) * train_data.n/batch.n - kl
+        
+        return elbo_fn
+    
+    # Compute KL divergence at inducing points, KL[q(u)||p(u)]: 
+    def prior_kl(self, params: Dict) -> Array:
+        mu = params["q"]["mu"]
+        sqrt = params["q"]["sqrt"]
+        nz = self.num_inducing
+
+        qu = distrax.MultivariateNormalTri(mu.squeeze(), sqrt)
+        
+        if not self.q.whiten:
+            z = params["q"]["inducing_inputs"]
+            mz = self.prior.mean_function(z, params["mean_function"])
+            Kzz = gram(self.prior.kernel, z, params["kernel"])
+            Kzz += I(nz) * self.jitter
+            Lz = cholesky(Kzz, lower=True)
+            pu = distrax.MultivariateNormalTri(mz.squeeze(), Lz)
+            
+        else:
+            pu = distrax.MultivariateNormalDiag(jnp.zeros(nz))
+        
+        return qu.kl_divergence(pu)
+
+
+    # Compute expectation, ∫ log(p(y|F)) q(F) dF, through Gauss-Hermite quadrature:
+    def variational_expectation(self, params: Dict, batch: Dataset) -> Array:
+        x, y = batch.X, batch.y
+
+        Fmu, Fvar = vmap(self.pred_moments, in_axes=(None, 0))(params, x[:,jnp.newaxis,:])
+
+        # Get log(p(y|F)) function for current likelihood parameter values:
+        def log_prob(F, y):
+            return self.likelihood.link_function(F, params["likelihood"]).log_prob(y)
+
+        return gauss_hermite_quadrature(log_prob, Fmu.squeeze(1), Fvar.squeeze(1), y=y)
+    
+    # Computes predictive moments for Gauss-Hermite quadrature:
+    def pred_moments(self, params: Dict, test_inputs: Array) -> Array:
+        mu = params["q"]["mu"]
+        sqrt = params["q"]["sqrt"]
+        
+        # Cholesky decomposition at inducing inputs:
+        z = params["q"]["inducing_inputs"]
+        nz = self.num_inducing
+        Kzz = gram(self.prior.kernel, z, params["kernel"])
+        Kzz += I(nz) * self.jitter
+        Lz = cholesky(Kzz)
+        
+        # Compute predictive moments:
+        t = test_inputs
+        Ktt = gram(self.prior.kernel, t, params["kernel"])
+        Kzt = cross_covariance(self.prior.kernel, z, t, params["kernel"])
+        M = solve_triangular(Lz, Kzt, lower=True)
+        Fcov = Ktt - jnp.matmul(M.T, M)
+    
+        if not self.q.whiten:
+            M = solve_triangular(Lz.T, M, lower=False)
+            mz = self.prior.mean(params)(z)
+            mu -= mz
```

Comment:
This conditional for computing `M` might be more cleanly handled as part of the variational family object.

Reply:
As above